### PR TITLE
Fix self-process identification for FreeBSD

### PR DESCRIPTION
**Pull request by josteink**
FreeBSD does not come with procfs enabled by default, and should use `sysctl()` for this purpose.

While it has similarities with NetBSD's implementation, there are a few subtle differences, which justifies leaving this implementation under its own guard.

It's also worth noting that on FreeBSD `sys/sysctl.h` MUST be present, which is unlike NetBSD. Therefore the `HAVE_SYS_SYSCTL_H` define is not checked for or used.

This commit fixes https://github.com/dotnet/coreclr/issues/6184.

This commit is based on the following commit from core-setup:
https://github.com/dotnet/core-setup/commit/d5ce08014a174b006a3b409b8bb93d003ae583a0

cc: @janvorli @jkotas

### CODE REVIEW

**Comment by josteink** [comment]
CI build fails due to failing Linux ARM tests, which seems to be the case for lots of PRs currently. Otherwise it appears good (errors caused by these changes should be in the first few percent of the build).

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp [hunk 1]
```diff
@@ -16,7 +16,11 @@
 #include <string>
 #include <string.h>
 #include <sys/stat.h>
-#ifdef HAVE_SYS_SYSCTL_H
```
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: A nit - I would slightly prefer keeping the `#ifdef HAVE_SYS_SYSCTL_H` intact and just add the other two headers under the __FreeBSD__ ifdef to have the include just once in there.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp [hunk 2]
```diff
@@ -75,6 +79,24 @@ bool GetEntrypointExecutableAbsolutePath(std::string& entrypointExecutable)
...
     }
+#elif defined (__FreeBSD__)
+    static const int name[] = {
+        CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1
+    };
+    char path[PATH_MAX];
+    size_t len;
+
+    len = sizeof(path);
+    if (sysctl(name, 4, path, &len, NULL, 0) == 0)
```
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: A nit - could you please use `__arraycount(name)` like the NetBSD version instead of 4?

Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: That was my first though as well, but it turns out `__arraycount()` is actually not by defined for FreeBSD unless you include `/usr/src/lib/libnetbsd/cdefs.h`. Which would mean we would need to probe for more stuff in the CMAKE-section which strictly speaking is not needed here. So I thought it was better without it. Any objections?

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 1 above)
Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Consider it done.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 2 above)
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Ah, ok. Makes sense.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 1 above)
Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: On a second note, for some reason the `HAVE_SYS_SYSCTL_H` cmake-checks fails, but in the process I discovered we also have a redundant `HAVE_SYSCTL` check which does *not* fail. And it used for the very same purpose: guarding `#include <sys/sysctl.h>` statements.

I don't know why we have 2 different ways to check for this same thing, but right now, if I were to have a choice, I would vote on eliminating the `HAVE_SYS_SYSCTL_H` check, as this should work on FreeBSD (and hopefully NetBSD too).

Sounds OK?

**Comment by josteink** [comment]
Pushed a new commit which eliminates the redundant `#include` statemnts, as well as the redundant `HAVE_SYS_SYSCTL_H` check in cmake's configuration process.

Let me know if think something more needs to be changed.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 1 above)
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: The problem is that on NetBSD, there is no sys/sysclt.h header, but the sysctl is present. So the include has to be guarded by the HAVE_SYS_SYSCTL_H. See the commit https://github.com/josteink/coreclr/commit/0eafec3e1065ba68c10f065d2713aaea0f75c160

Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Ah, it was your commit :-) Interesting.

Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Nope. Those were done by @jasonwilliams200OK. I don't have anything to do with those, and I'll just revert the changes I did to make sure I don't ruin anyone else changes :)

**Comment by josteink** [comment]
Pushed a new commit, with only FreeBSD specific changes, and no changes to the NetBSD/Alpine Linux bits. Hopefully this should make everyone happy.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 1 above)
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: You can try to let cmake generate preprocessed src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp and see where the sysctl function comes from. Just go to the `bin/obj/Linux.x64.Debug/src/coreclr/hosts/unixcoreruncommon` (replace Linux.x64.Debug with whatever you have) and run `make coreruncommon.cpp.i`. Then look at the generated file to see where the sysctl came from.

Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Looking into `CMakeErrors.log` instead, I found some more details, which explains the problem:

````
Dermining if files sys/sysctl.h exist failed with the following output:
Change Dir: /home/jostein/build/coreclr/bin/obj/FreeBSD.x64.Debug/CMakeFiles/CMakeTmp

Run Build Command:"/usr/local/bin/gmake" "cmTC_688ff/fast"
/usr/local/bin/gmake -f CMakeFiles/cmTC_688ff.dir/build.make CMakeFiles/cmTC_688ff.dir/build
gmake[1]: Entering directory '/home/jostein/build/coreclr/bin/obj/FreeBSD.x64.Debug/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_688ff.dir/CheckIncludeFiles.c.o
/usr/local/bin/clang37  -D_FILE_OFFSET_BITS=64 -Wall -std=c11  -I/usr/local/include    -o CMakeFiles/cmTC_688ff.dir/CheckIncludeFiles.c.o   -c /home/jostein/build/coreclr/bin/obj/FreeBSD.x64.Debug/CMakeFiles/CMakeTmp/CheckIncludeFiles.c
In file included from /home/jostein/build/coreclr/bin/obj/FreeBSD.x64.Debug/CMakeFiles/CMakeTmp/CheckIncludeFiles.c:2:
/usr/include/sys/sysctl.h:804:40: error: unknown type name 'size_t'
int     sysctl(const int *, u_int, void *, size_t *, const void *, size_t);
                                           ^
/usr/include/sys/sysctl.h:805:40: error: unknown type name 'size_t'
int     sysctlbyname(const char *, void *, size_t *, const void *, size_t);
                                           ^
/usr/include/sys/sysctl.h:806:42: error: unknown type name 'size_t'
int     sysctlnametomib(const char *, int *, size_t *);
````

Basically (as can be seen in my commit), on FreeBSD you cannot simply include `sys/sysctl.h` without the proper prerequisites from `sys/types.h`. This is what's causing the cmake-check to fail.

For now, I`d say that the changes required to make this test pass is not worth it. If we have more `sys/sysctl.h` related issued in the future, we can reconsider. Sounds OK?

Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Sounds ok for now. Thanks for tracking down the culprit. While I would like to eventually fix the test, I'd like to make sure it fixes the problem on the Alpine Linux and others that caused the addition of the HAVE_SYS_SYSCTL_H. I'll create an issue for it.

Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Um, please ignore that part about fixing it for Alpine etc.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 2 above)
Comment (adityamandaleeka) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Minor nit: can you use `nullptr` here to match the rest of the file (there is one other occurrence of `NULL` but the rest is all `nullptr`?

**Comment by adityamandaleeka** [comment]
LGTM module minor comment.

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 2 above)
Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Done and pushed new commit.

**Comment by josteink** [comment]
OK. With no other objections, I guess this should now be good to merge :)

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp (hunk 1 above)
Comment (jasonwilliams200OK) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: This patch in cmake file fixes the problem for both OSes:

```diff
-check_include_file(sys/sysctl.h HAVE_SYS_SYSCTL_H)
+check_include_file(sys/sysctl.h HAVE_SYS_SYSCTL_H)
+# Freebsd fails with this test
+if (NOT HAVE_SYS_SYSCTL_H)
+   CHECK_C_SOURCE_COMPILES("#include <sys/types.h>\n#include <sys/sysctl.h>\nint main(void) { };" HAVE_SYS_SYSCTL_H_TMP)
+    if ( HAVE_SYS_SYSCTL_H_TMP )
+      SET(HAVE_SYS_SYSCTL_H 1)
+    endif ( HAVE_SYS_SYSCTL_H_TMP )
+endif (NOT HAVE_SYS_SYSCTL_H)
```

Comment (josteink) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: That's actually very good, and I'll rather adopt that instead. Thanks for the help!

**Comment by josteink** [comment]
While I was able to get @jasonwilliams200OK's cmake-patch pass, I couldn't get the result properly exported, result being that compilation still failed without checking for FreeBSD as a platform.

I can see a few "hacky" workarounds, but for now I guess I'll revert to my old conclusion: getting this particular cmake-business working is more effort than it's seemingly worth.

If that's OK with everyone else, feel free to merge as is. If not, please let me know.

**Comment by janvorli** [comment]
@josteink can you please be more specific on what you mean by "I couldn't get the result properly exported"?

**Comment by josteink** [comment]
The result of the first include-test was written to CMake's cache before our overridden value. This made the overall result (and config.h) the same as before. Trying to apply a few different approaches (like replacing the original test with the patch itself) resulted in the cache to look like it should (as did also the generated config.h), but for some reason the guard still didn't trigger.

There's obviously something going wrong with the configuration process, but it seems unrelated to my patch as such, and I don't really have time to look deeper into it now...

**Comment by janvorli** [comment]
Thanks for the explanation. Let's keep your change as it is and fix that problem later.
LGTM

**Comment by jasonwilliams200OK** [comment]
@josteink, did you get that result by using that patch as is, i.e. `HAVE_SYS_SYSCTL_H` to first check header then if it fails use `HAVE_SYS_SYSCTL_H_TMP` (temp variable) to store the ret value of `CHECK_C_SOURCE_COMPILES` then set `HAVE_SYS_SYSCTL_H` if it does? If you dropped that `_TMP` thingy then it will fail the way you described?

**Comment by josteink** [comment]
This PR is IMO already too littered with comments. To reduce further noise, if you want a full explanation, I may be able to provide that on gitter later on. OK?

@janvorli: OK to merge?

**Comment by jasonwilliams200OK** [comment]
> OK?

umm, I don't get this same rush in every PR.

> if you want a full explanation

No i don't want full "explanation". You mentioned me in the PR, I created a patch, tested on two operating systems Alpine and FreeBSD, and proposed. It failed on your machine, I asked "did you apply the same patch or modified it", to figure out if I missed something. A yes or no answer would have suffice. 

> This PR is IMO already too littered with comments. To reduce further noise

Lots of comment in PR is not a bad idea as long as they are on-topic. GitHub is a place of collaboration and discussion-driven contributions make the software better. There is no need, and no place, for attitude. OK?

**Comment by josteink** [comment]
I didn't mean to have any attitude. I'm very busy and thought you wanted me to outline in detail what I had attempted, which I did not want. If all you ask for is a yes or no answer, the answer is yes.

**Comment by mkborg** [comment]
@dotnet-bot test Linux ARM Emulator Cross Release Build

**Pull request merged**